### ISSUE TITLE
DIG DUG original disk fails to load

### ISSUE THREAD

**Issue by Jusalak** [GideonZ/1541ultimate#304]
Original game disk of DIG DUG (c) 1983 Namco, with V-MAX! fails to load. Virtual drives A and B were disabled, game loading attempt from real 1541-II resulted in a crash.

On a real C64C with the same 1541-II drive the game loaded successfully on the first try.

**Comment by markusC64** [comment]
Well, I tried exactly that disk on my Ultimate 64 with my external 1541 II. It does work!

The interesting part might be what wo check., No real idea, since even the IEC device makes no difference on my U64.

**Comment by markusC64** [comment]
I've tested that disk on my C128 with a 1541 II. It shows that it hangs early in loading if some fast loader kernal is active. Original CBM kernal has to be used.

**Comment by Jusalak** [comment]
The disk has "C64/128" so it is expected to run on 1571, so I tried it on 1571. A real C64C + 1571 loaded it fine.

U64 + 1571 again failed. A stock CBM Kernal has been used. Tried also restoring default settings, made no difference.

The loading screen has "ThunderMountain" logo on it (and that printed on the disk as well). U64 gets this far but then crashes and displays a garbled screen for a second, then displays the C64 startup screen.

**Comment by markusC64** [comment]
Well, there is always the possibility that there are different versions of that game. Your descriptions sounds different.
Without a way to reproduce the issue, it cannot be analyzed.

**Comment by Jusalak** [comment]
Yes, C128 was released in 1985 so the disk I have can hardly be earlier than that - yet the C64 version of DIG DUG was released was released a few years prior. So this disk is of a later batch with presumably an improved loader/protection.

BTW, there is an improved version of the game which fixes a bug in the original game when it gets stuck after too many monsters are killed by a single rock.

https://csdb.dk/release/?id=225869&show=notes

**Comment by markusC64** [comment]
That's interesting - Now using the right disk in my 1541 II, I have to confirm it.

**Comment by Jusalak** [comment]
Revolution V V1.5 reports V-MAX version 1. https://csdb.dk/release/?id=134573. Could not copy since version 1 is not supported.

This is what c64preservation.com has to say about version 1: "This is used in some Activision games and is mostly easily copied and works in VICE. A couple titles have an extra byte counting protection check that still fails. It uses normal CBM DOS sectors. "

https://c64preservation.com/dp.php?pg=vmax

**Comment by Zibri** [comment]
> Original game disk of DIG DUG (c) 1983 Namco, with V-MAX! fails to load. Virtual drives A and B were disabled, game loading attempt from real 1541-II resulted in a crash.
> 
> On a real C64C with the same 1541-II drive the game loaded successfully on the first try.
> 
> U64 with latest firmware.

Can you post a link to the original disk?

**Comment by Jusalak** [comment]
That link does not work. Please share a working link.

**Comment by Zibri** [comment]
> Revolution V V1.5 reports V-MAX version 1. https://csdb.dk/release/?id=134573. Could not copy since version 1 is not supported.
> 
> This is what c64preservation.com has to say about version 1: "This is used in some Activision games and is mostly easily copied and works in VICE. A couple titles have an extra byte counting protection check that still fails. It uses normal CBM DOS sectors. "
> 
> https://c64preservation.com/dp.php?pg=vmax
> 
> Edit: Nice if the disk image could be shared, I would like to try it on other emulators and also on a Pi1541 device on a C64.

[You can find everything here.](https://archive.org/details/C64_Preservation_Project_10th_Anniversary_Collection)

**Comment by Jusalak** [comment]
Enabling REU makes it crash on VICE as well.

**Comment by Jusalak** [comment]
I looked a bit why emulators (Vice and Z64K) failed with REU (since I cannot debug U64).
The JSR $1044 instruction at $1016 seems to be the key. They get this far but they don't seem to get past this.

**Comment by johnlivdahl** [comment]
I'll try to post a direct like to the VMAX version that fails on U64: https://archive.org/download/C64_Preservation_Project_10th_Anniversary_Collection/C64_Preservation_Project_10th_Anniversary_Collection_G64.zip/c64pp-g64-zip%2Fd%2Fdig_dug%5Bthunder_mtn_198x%5D%28vmax2%29%28%21%29.zip

The other 2 G64 versions both load fine.  The linked version does load fine on VICE 3.1 even with Jiffy Dos kernal and drive ROM enabled.

**Comment by Jusalak** [comment]
I looked at the code a bit more closely. After waiting for the raster it then accesses the presumably empty I/O range $DE00-$DFFF. Incompatibility with the REU also gets an explanation since it messes up with the REU I/O range. My guess is that U64 does not emulate correctly something related to that I/O range.
Drive emulation does not seem to be the cause since loading failed with real physical drive and floppy. I think a real C64/128 with UII will run it correctly.

![code](https://user-images.githubusercontent.com/110191335/235291465-2b5aad73-c7ef-4617-822e-cf2b68482445.PNG)

**Comment by Jusalak** [comment]
I further tried modifying the code:

```
108c LDA $DE00,X

10a0 LDA $DE00,X
```

With these modifications the presence of REU no longer prevents the program from running.
The accurate emulation of this I/O range is required to run the loader successfully.

**Comment by johnlivdahl** [comment]
That code might be freezer cartridge detention code that I thought Gideon may have fixed for an issue a few years back.  Something like if no IO1 or IO2 range device is plugged in the code expects to see the VIC data still on the bus from the last VIC read.

**Comment by GideonZ** [comment]
This is broken since the introduction of the turbo mode.
There is another ticket that describes this problem with another title,
together with the proposed solution.

**Comment by johnlivdahl** [comment]
Here is the link to the other issue: https://github.com/GideonZ/1541ultimate/issues/205

**Comment by Jusalak** [comment]
The patch seems to be
```
105a eor $115f,x
...
1068 lda $1172,x
...
107c lda $1172,x
...
108c lda $1186,x
...
10a0 lda $1186,x
```
If U64 had memory poke functionality, this could be done easily (during loading). But since it does not seem to have such, it cannot be done.

**Comment by GideonZ** [comment]
It actually does have this functionality.
In 3.11 you can also do it through the API and set a breakpoint.

**Comment by Grrrolf** [comment]
If this is indeed a duplicate issue, then please close this issue and add your solution resolving the problem in the original issue. There is no need to keep duplicate issues open.

**Comment by GideonZ** [comment]
The solution is already there, Rolf.. It just needs to be implemented and
verified.

**Comment by Jusalak** [comment]
Where is the memory poke function, I don't see it anywhere?

**Comment by GideonZ** [comment]
For now you can use the soon-to-be-deprecated socket connection: TCP port
64, and send the sequence 06 FF <low addr> <high addr> [<data>]*

From 3.11, you can do a HTTP request:

[PUT] http://<ipaddr>/v1/machine:writemem?address=<hexaddr>&data=<hexdata>
(up to 128 bytes)
[POST] http://<ipaddr>/v1/machine:writemem?address=<hexaddr> with the data
in the body as a binary attachment

There is also a poke command when the firmware is compiled with the
DEVELOPER flag; accessible from the F5 menu. But that's single byte per
menu access.

**Comment by Jusalak** [comment]
Actually it suffices to modify only two bytes, or one line of code

`
105a 5d 5f 11 eor $115f,x
`

Those two EOR instructions could be replaced with NOPs, but that takes six bytes. Either way, at the end of the loop the accumulator will have zero, and the following BEQ instuction will jump to a RTS instruction, and so the subroutine always passes.

```
$105b,$5f (decimal 4187,95)
$105c,$11 (decimal 4188,17)
```
My humble opinion is that memory poking should be enabled by default.

**Comment by GameTimeWarp** [comment]
Any chance of getting this issue fixed before the next firmware update?

**Comment by GameTimeWarp** [comment]
These V-Max games still seem to be broken on the latest 3.12 firmware. Any chance of slipping the fix in for the next update @GideonZ ?

**Comment by Jumpman64** [comment]
"Dig Dug Thunder Mountain 198x (vmax2)(!)" is working now with firmware (V1.4F) 3.15 
also see https://github.com/GideonZ/1541ultimate/issues/205
Looks like every Copy Protection, that refuses to load or run a game when a cartridge is pluged in, should work now.
